### PR TITLE
add scan over layers section

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

Adds a `Scan Over Layers` section to the Haiku Migration Guide.

Preview: https://flax--3195.org.readthedocs.build/en/3195/guides/haiku_migration_guide.html#scan-over-layers

### CODE REVIEW

File: docs/guides/haiku_migration_guide.rst [hunk 1]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
 
 The only notable change with respect to the examples in the previous sections is that
 this time around we used ``hk.without_apply_rng`` in Haiku so we didn't have to
-pass the ``rng`` argument as ``None`` to the ``apply`` method.
\ No newline at end of file
+pass the ``rng`` argument as ``None`` to the ``apply`` method.
+
+Scan Over Layers
+----------------
+One very important application of ``scan`` is apply a sequence of layers iteratively
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:720]: There's `jax.lax.scan` and `jax.linen.scan`, so let's use full API names. Also a small grammar fix.

```suggestion
One very important application of ``flax.linen.scan`` is to apply a sequence of layers iteratively
```

File: docs/guides/haiku_migration_guide.rst [hunk 2]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
...
+over an input, passing the output of each layer as the input to the next layer. This
+very useful to reduce compilation time for big models. As an example we will create
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:722]:
```suggestion
is very useful to reduce compilation time for big models. As an example we will create
```

File: docs/guides/haiku_migration_guide.rst [hunk 3]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
...
+a simple ``Block`` Module, and then use it inside an ``MLP`` Module that will apply
+the ``Block`` Module ``num_layers`` times.
+
+In Haiku, we define the ``Block`` Module as usual, and then inside ``MLP`` we will
+use ``hk.experimental.layer_stack`` over a ``stack_block`` function to create a stack
+of ``Block`` Modules. In Flax, the definition of ``Block`` is a little different,
+``__call__`` will accept and return a second dummy input/output that in both cases will
+be ``None``. In ``MLP``, we will use ``nn.scan`` as in the previous example, but
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:730]:
```suggestion
be ``None``. In ``MLP``, we will use ``flax.linen.scan`` as in the previous example. But
```

File: docs/guides/haiku_migration_guide.rst [hunk 4]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
...
+by setting ``split_rngs={'params': True}`` and ``variable_axes={'params': 0}``
+we are telling ``nn.scan`` create different parameters for each step and slice the
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:732]:
```suggestion
we are instructing ``flax.linen.scan`` to create different parameters for each step and slice the
```

File: docs/guides/haiku_migration_guide.rst [hunk 5]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
...
+    def __call__(self, x, training: bool):
+      ScanBlock = nn.scan(
+        Block, variable_axes={'params': 0}, split_rngs={'params': True},
+        length=self.num_layers)
+
+      y, _ = ScanBlock(self.features, training)(x, None)
+      return y
+
+Notice how in Flax we pass ``None`` as the second argument to ``ScanBlock`` and ignore
+its second output, these represent the inputs/outputs per-step but they are ``None``
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:794]:
```suggestion
its second output. These represent the inputs/outputs per-step but they are ``None``
```

File: docs/guides/haiku_migration_guide.rst [hunk 6]
```diff
@@ -713,4 +713,152 @@ Finally, let's quickly view how the ``RNN`` Module would be used in both Haiku a
 
 The only notable change with respect to the examples in the previous sections is that
 this time around we used ``hk.without_apply_rng`` in Haiku so we didn't have to
-pass the ``rng`` argument as ``None`` to the ``apply`` method.
\ No newline at end of file
+pass the ``rng`` argument as ``None`` to the ``apply`` method.
+
+Scan Over Layers
```
Comment (8bitmp3) [inline docs/guides/haiku_migration_guide.rst:718]:
```suggestion
Scan over layers
```

We can implement other style guide-related changes to the rest of the doc in another PR.

**Comment by 8bitmp3** [comment]
@LenaMartens LMKWYT

**Comment by cgarciae** [comment]
@8bitmp3 as @levskaya mentioned, we should not publicize the `flax.linen` namespace explicitly so much in case we change it to `nn` (`flax.nn`) in the near future.

**Review by IvyZX — APPROVED** [review]
Approved for the code content - thank you!

One other thing could be added to the doc is to showcase how Haiku may be able to run model code and even transform with functions, whereas in Flax a module is the recommended way to go, esp. to ensure lifted transform + partitioning works.

**Pull request merged**